### PR TITLE
feat: Added allowed routes to TradeParams

### PR DESCRIPTION
**Pull request by peterMangrove**

### CODE REVIEW

File: packages/mangrove.js/src/util/trade.ts [hunk 1]
```diff
@@ -153,18 +153,27 @@ class Trade {
...
+            market: market,
+            fillOrKill: params.fillOrKill ? params.fillOrKill : false,
+          },
+          overrides
+        );
+      } else {
+        throw new Error(
+          "Given parameters matches a mangroveOrder, but MangroveOrder is not in the params.allowedOrderRoutes"
+        );
+      }
```
Comment (espendk) [inline packages/mangrove.js/src/util/trade.ts:176]: This can be reverted I think

File: packages/mangrove.js/src/util/trade.ts [hunk 2]
```diff
@@ -184,16 +193,25 @@ class Trade {
...
+              fillWants: fillWants,
+              market,
+            },
+            overrides
+          );
+        } else {
+          throw new Error(
+            "Given parameters matches a marketOrder, but Mangrove is not in the params.allowedOrderRoutes"
+          );
+        }
```
Comment (espendk) [inline packages/mangrove.js/src/util/trade.ts:214]: This can be reverted I think

**Review by espendk — CHANGES_REQUESTED** [review]
I think the `OrderRoute` solution will take more work to get good, so I suggest we use a simple flag to force routing of orders to `MangroveOrder` instead, see the code comments.

File: packages/mangrove.js/src/util/trade.ts (hunk 1 above)
Comment (peterMangrove) [reply packages/mangrove.js/src/util/trade.ts]: Done

File: packages/mangrove.js/src/util/trade.ts (hunk 2 above)
Comment (peterMangrove) [reply packages/mangrove.js/src/util/trade.ts]: Done

File: packages/mangrove.js/src/util/trade.ts [hunk 3]
```diff
@@ -153,18 +153,27 @@ class Trade {
     const restingOrderParams =
       "restingOrder" in params ? params.restingOrder : null;
     if (params.fillOrKill || restingOrderParams) {
```
[the remark this answers is not in the document]
Comment (peterMangrove) [reply packages/mangrove.js/src/util/trade.ts]: Done

[the remark this answers is not in the document]
Comment (peterMangrove) [reply packages/mangrove.js/src/util/trade.ts]: Done

File: packages/mangrove.js/src/market.ts [hunk 4]
```diff
@@ -68,7 +68,10 @@ namespace Market {
     | ({ name: "OfferRetract" } & TCM.OfferRetractEvent)
     | ({ name: "SetGasbase" } & TCM.SetGasbaseEvent);
 
+  export type OrderRoute = "Mangrove" | "MangroveOrder";
+
   export type TradeParams = {
+    allowedOrderRoutes?: OrderRoute[];
```
[the remark this answers is not in the document]
Comment (peterMangrove) [reply packages/mangrove.js/src/market.ts]: Done

File: packages/mangrove.js/src/market.ts [hunk 5]
```diff
@@ -68,7 +68,10 @@ namespace Market {
     | ({ name: "OfferRetract" } & TCM.OfferRetractEvent)
     | ({ name: "SetGasbase" } & TCM.SetGasbaseEvent);
 
+  export type OrderRoute = "Mangrove" | "MangroveOrder";
+
```
[the remark this answers is not in the document]
Comment (peterMangrove) [reply packages/mangrove.js/src/market.ts:72]: Done

File: packages/mangrove.js/src/util/trade.ts [hunk 6]
```diff
@@ -152,7 +152,11 @@ class Trade {
         : this.getParamsForSell(params, market.base, market.quote);
     const restingOrderParams =
       "restingOrder" in params ? params.restingOrder : null;
-    if (params.fillOrKill || restingOrderParams) {
+    if (
+      params.fillOrKill ||
+      restingOrderParams ||
+      (params.forceRoutingToMangroveOrder ?? false)
```
Comment (espendk) [inline packages/mangrove.js/src/util/trade.ts:158]: `fillOrKill` and `forceRoutingToMangroveOrder` are both optional booleans, so they should be treated the same here. I think the idiomatic way would be to use `!!optionalBoolean`, ie `!!params.fillOrKill` and `!!params.forceRoutingToMangroveOrder`.

**Review by espendk — CHANGES_REQUESTED** [review]

**Review by espendk — APPROVED** [review]

**Pull request merged**